proposals/semi-auto-properties.md

```diff
@@ -10,6 +10,257 @@ Two common scenarios are that you want to apply a constraint on the setter, ensu
+
+Here's an example of where this is useful. The `field` keyword will find a lot of its use with view models because of the neat solution it brings for the `INotifyPropertyChanged` pattern. View model property setters are likely to be databound to UI and likely to cause change tracking or trigger other behaviors. The following code needs to initialize the default value of `IsActive` without setting `HasPendingChanges` to `true`:
+
+```cs
+using System.Runtime.CompilerServices;
+
+class SomeViewModel
+{
+    public bool HasPendingChanges { get; private set; }
+
+    public bool IsActive { get; set => Set(ref field, value); } = true;
+
+    private bool Set<T>(ref T location, T value)
+    {
+        if (RuntimeHelpers.Equals(location, value)) return false;
+        location = value;
+        HasPendingChanges = true;
+        return true;
+    }
+}
+```
+
+This difference in behavior between a property initializer and assigning from the constructor can also be seen with virtual auto-properties in previous versions of the language:
+
+```cs
+using System;
+
+// Nothing is printed; the property initializer is not
+// equivalent to `this.IsActive = true`.
+_ = new Derived();
+
+class Base
+{
+    public virtual bool IsActive { get; set; } = true;
+}
+
+class Derived : Base
+{
+    public override bool IsActive
+    {
+        get => base.IsActive;
+        set
+        {
+            base.IsActive = value;
+            Console.WriteLine("This will not be reached");
+        }
+    }
+}
+```
+
+### Constructor assignment
+
+As with existing auto-properties, assignment in the constructor calls the setter if it exists, and if there is no setter it falls back to directly assigning to the backing field.
+
+```cs
+class C
+{
+    public C()
+    {
+        P1 = 1; // Assigns P1's backing field directly
+        P2 = 2; // Assigns P2's backing field directly
+        P3 = 3; // Calls P3's setter
+        P4 = 4; // Calls P4's setter
+    }
+
+    public int P1 => field;
+    public int P2 { get => field; }
+    public int P4 { get => field; set => field = value; }
+    public int P3 { get => field; set; }
+}
+```
+
+### Definite assignment in structs
+
+A semi-auto property will be treated as a regular auto property for the purposes of calculating default backing field initialization if its setter is automatically implemented, or if it does not have a setter ([LDM decision](https://github.com/dotnet/csharplang/blob/main/meetings/2022/LDM-2022-03-02.md#property-assignment-in-structs)).
+
+Default initialize a struct when calling a manually implemented semi-auto property setter, and issue a warning when doing so, like a regular property setter ([LDM decision](https://github.com/dotnet/csharplang/blob/main/meetings/2022/LDM-2022-05-02.md#definite-assignment-of-manually-implemented-setters)).
```

Comment:
might be good to have examples of these cases here.

Reply:
@RikkiGibson Would this be the right pair of examples?

```cs
public struct C
{
    public C()
    {
        _ = P; // No warning
    }

    public int P { get => field; }
}
```

```cs
public struct C
{
    public C()
    {
        P = 5; // No warning
    }

    public int P { get => field; set => field = value; }
}
```

Reply:
Change `No warning` to `Disabled-by-default warning` and you got it.

Reply:
Hmm. With disabled-by-default warning, is it a separate warning from the one that I'm about to say will not be shown based on the LDM decision link? Full thing I'm about to say:

> Default-initialize a struct when calling a manually implemented setter of a property which uses an automatic backing field, and issue a warning when doing so, like a regular property setter ([LDM decision](https://github.com/dotnet/csharplang/blob/main/meetings/2022/LDM-2022-05-02.md#definite-assignment-of-manually-implemented-setters)). That enables the following code to compile without warning "CS9020: The 'this' object is read before all of its fields have been assigned, causing preceding implicit assignments of 'default' to non-explicitly assigned fields":
> 
> ```cs
> public struct S
> {
>     public S()
>     {
>         P = 5; // Only a disabled-by-default warning
>     }
> 
>     public int P { get => field; set => field = value; }
> }
> ```

Similar question for the other code block.

Reply:
I think I'm confused by the language here. It seems like the spec is saying "default-initialize, and report a warning, so that we don't have to report this other particular warning."

I think the spec shouldn't concern itself with the specific ID or message of the warning, it should just say that the warning is reported and whether the warning is enabled by default.

I would just say that the field(s) denoted by the `field` keyword are subject to default-initialization and disabled-by-default warnings under the same conditions as any other struct fields--such as when a manually-implemented property setter is called before the fields are assigned--even though such fields are not directly referenceable in the constructor. Then include the samples.

Reply:
Do you feel that the linked LDM decisions are explained in sufficient detail here, or should I add more?

> Even though they can't be referenced in the constructor, backing fields denoted by the `field` keyword are subject to default-initialization and disabled-by-default warnings under the same conditions as any other struct fields. ([LDM decision 1](https://github.com/dotnet/csharplang/blob/main/meetings/2022/LDM-2022-03-02.md#property-assignment-in-structs), [LDM decision 2](https://github.com/dotnet/csharplang/blob/main/meetings/2022/LDM-2022-05-02.md#definite-assignment-of-manually-implemented-setters)). For example:
> 
> ```cs
> public struct S
> {
>     public S()
>     {
>         _ = P1; // Disabled-by-default warning
>         P2 = 5; // Disabled-by-default warning
>     }
> 
>     public int P1 { get => field; }
>     public int P2 { get => field; set => field = value; }
> }
> ```

Reply:
Updated. I did confirm that the warning is only present up to one time at a time in the constructor.